pkg/git/libgit2/transport.go

```diff
@@ -173,13 +175,28 @@ func parseKnownHosts(s string) ([]knownKey, error) {
 	return knownHosts, nil
 }
 
-func (k knownKey) matches(host string, key []byte) bool {
+func (k knownKey) matches(host string, hostkey git2go.HostkeyCertificate) bool {
 	if !containsHost(k.hosts, host) {
 		return false
 	}
 
-	hash := sha1.Sum([]byte(k.key.Marshal()))
-	if bytes.Compare(hash[:], key) != 0 {
+	var hash []byte
+	var key []byte
+	switch hostkey.Kind {
+	case git2go.HostkeyMD5:
+		sum := md5.Sum(k.key.Marshal())
+		hash = sum[:]
+		key = hostkey.HashMD5[:]
+	case git2go.HostkeySHA1:
+		sum := sha1.Sum(k.key.Marshal())
+		hash = sum[:]
+		key = hostkey.HashSHA1[:]
+	case git2go.HostkeySHA256:
+		sum := sha256.Sum256(k.key.Marshal())
+		hash = sum[:]
+		key = hostkey.HashSHA256[:]
+	}
+	if bytes.Compare(hash, key) != 0 {
```

Comment:
It may be a bit prettier to do something like the following (lengthwise it does not make much of a difference):

```suggestion
	var fingerprint []byte
	var hasher hash.Hash
	switch hostkey.Kind {
	case git2go.HostkeyMD5:
		fingerprint = hostkey.HashMD5[:]
		hasher = md5.New()
	case git2go.HostkeySHA1:
		fingerprint = hostkey.HashSHA1[:]
		hasher = sha1.New()
	case git2go.HostkeySHA256:
		fingerprint = hostkey.HashSHA256[:]
		hasher = sha256.New()
	default:
		return false
	}
	hasher.Write(k.key.Marshal())
	if bytes.Compare(hasher.Sum(nil), fingerprint) != 0 {
		return false
	}
```